gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java

```diff
@@ -1005,6 +1005,162 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
     }
   }
 
+  private void checkRecoveredIndex(ArrayList<GemFireContainer> uninitializedContainers,
+      final LogWriter logger, boolean throwErrorOnMismatch) {
+    try {
+      for (GemFireContainer container : uninitializedContainers) {
+        LocalRegion region = container.getRegion();
+        DataPolicy dp = region.getDataPolicy();
+        if (dp == DataPolicy.PERSISTENT_PARTITION || dp == DataPolicy.PERSISTENT_REPLICATE) {
+          GfxdIndexManager gim = (GfxdIndexManager)region.getIndexUpdater();
+          if (gim == null || container.isGlobalIndex()) continue;
+          int localRegionSz = getLocalRegionSz(region, dp, logger, false, throwErrorOnMismatch);
+          List<GemFireContainer> allIndexes = gim.getAllIndexes();
+          for (GemFireContainer c : allIndexes) {
+            if (c.isLocalIndex()) {
+              if (c.getIndexSize() != localRegionSz) {
+                if (!throwErrorOnMismatch) {
+                  logger.warning("checkRecoveredIndex: for table: " + region.getName() + " " +
+                      "number of local entries = " + localRegionSz + " and number of " +
+                      "index entries in the index: " + c.getName() + " = " + c.getIndexSize());
+                }
+                else {
+                  logger.error("checkRecoveredIndex: for table: " + region.getName() + " " +
+                      "number of local entries = " + localRegionSz + " and number of " +
+                      "index entries in the index: " + c.getName() + " = " + c.getIndexSize());
+                  dumpIndexAndRegion(region, dp, c, logger);
+                  throw new IllegalStateException("Table data and indexes are not reconciling." +
+                      "Probably need to revoke the disk store");
+                }
+                logger.info("FabricDatabase: index and region out of sync. Recreating the indexes");
+                // First clear all the indexes
+                clearAllIndexes(uninitializedContainers);
+                recreateAllLocalIndexes(logger);
+                checkRecoveredIndex(uninitializedContainers, logger, true);
+              } else {
```

Comment:
Since we are only using count match to ascertain whether index built might be correct or not and not looking in the exact value so I thought in case of discrepancy which might have even side effects, let's be a bit conservative and rebuild so that irf and crf are guaranteed to be good both in terms of number and exact value.